docs/classads/classad-mechanism.rst

```diff
@@ -914,7 +914,41 @@ Optional parameters are given within square brackets.
     are omitted from the string. For example, ``seconds`` of 67 becomes
     "1:07". A second example, ``seconds`` of 1472523 = 17\*24\*60\*60 +
     1\*60\*60 + 2\*60 + 3, results in the string "17+1:02:03".
-    
+
+    :index:`evalInEachContext()<single: evalInEachContext(); ClassAd functions>`
+``String evalInEachContext(Expression expr, List contexts)``
+    This function evaluates its first argument as an expression in the context of
+    each Classad in the second argument and returns a list that is the result of
+    each evaluation. The first argument should be an expression. 
+    If the second argument does not evaluate to a list of ClassAds, ``ERROR`` is returned.
```

Comment:
(I really like the examples here.  I'd love to add an example for `evalInEachContext(gpu_requirements, ...)`, but I'm not sure that can be done compactly enough.)

This documentation doesn't mention the special handling of attribute references, and I think it probably should; the behavior is clearly better than being consistent about it, but I can see it being very mysterious without an explanation.

Maybe something like the following, assuming I actually understand all the cases?

This function returns a list of evaluation results, one for each ClassAd in the second argument, and in the same order. (If second argument does not evaluate to a list of ClassAds, ``ERROR`` is returned.)  The expression is specified by the first argument.  If the first argument is an expression literal, that expression is evaluated in each context.  If the first argument is just the name of an attribute, and it is defined in ``MY`` or ``TARGET`` (looked up in the usual way), the value of that attribute reference is evaluated in each context.  If the attribute reference is not defined in ``MY`` or ``TARGET``, the attribute reference will be looked up in each context and evaluated there.  This allows you to use this function to get `attribute_name` from each of the context ads, like `contexts.attribute_name`.

I think the documentation for `countMatches()` could just reference the documentation for `evalInEachContext()` where it needs to talk about how the first argument behaves.

Unless I missed something, the language about the second argument should be the same for both functions (e.g., "a list of ClassAds, a list of attribute references to ClassAds, or evaluate to a list of ClassAds").

Reply:
countMatches is more tolerant of variations in the second argument, so I think it might be confusing for one to reference the other.

Reply:
TJ made the further point that more-specific documentation also limits our room for future maneuver.  I still feel like this documentation will need improvement, but I'm also willing to let it go until somebody actually complains.